// react-css-animate/css-easing.js
'use strict';

exports.__esModule = true;
// Cubic
var easeInCubic = exports.easeInCubic = 'cubic-bezier(0.550, 0.055, 0.675, 0.190)';
var easeOutCubic = exports.easeOutCubic = 'cubic-bezier(0.215, 0.610, 0.355, 1.000)';
var easeInOutCubic = exports.easeInOutCubic = 'cubic-bezier(0.645, 0.045, 0.355, 1.000)';

// Circ
var easeInCirc = exports.easeInCirc = 'cubic-bezier(0.600, 0.040, 0.980, 0.335)';
var easeOutCirc = exports.easeOutCirc = 'cubic-bezier(0.075, 0.820, 0.165, 1.000)';
var easeInOutCirc = exports.easeInOutCirc = 'cubic-bezier(0.785, 0.135, 0.150, 0.860)';

// Expo
var easeInExpo = exports.easeInExpo = 'cubic-bezier(0.950, 0.050, 0.795, 0.035)';
var easeOutExpo = exports.easeOutExpo = 'cubic-bezier(0.190, 1.000, 0.220, 1.000)';
var easeInOutExpo = exports.easeInOutExpo = 'cubic-bezier(1.000, 0.000, 0.000, 1.000)';

// Quad
var easeInQuad = exports.easeInQuad = 'cubic-bezier(0.550, 0.085, 0.680, 0.530)';
var easeOutQuad = exports.easeOutQuad = 'cubic-bezier(0.250, 0.460, 0.450, 0.940)';
var easeInOutQuad = exports.easeInOutQuad = 'cubic-bezier(0.455, 0.030, 0.515, 0.955)';

// Quart
var easeInQuart = exports.easeInQuart = 'cubic-bezier(0.895, 0.030, 0.685, 0.220)';
var easeOutQuart = exports.easeOutQuart = 'cubic-bezier(0.165, 0.840, 0.440, 1.000)';
var easeInOutQuart = exports.easeInOutQuart = 'cubic-bezier(0.770, 0.000, 0.175, 1.000)';

// Quint
var easeInQuint = exports.easeInQuint = 'cubic-bezier(0.755, 0.050, 0.855, 0.060)';
var easeOutQuint = exports.easeOutQuint = 'cubic-bezier(0.230, 1.000, 0.320, 1.000)';
var easeInOutQuint = exports.easeInOutQuint = 'cubic-bezier(0.860, 0.000, 0.070, 1.000)';

// Sine
var easeInSine = exports.easeInSine = 'cubic-bezier(0.470, 0.000, 0.745, 0.715)';
var easeOutSine = exports.easeOutSine = 'cubic-bezier(0.390, 0.575, 0.565, 1.000)';
var easeInOutSine = exports.easeInOutSine = 'cubic-bezier(0.445, 0.050, 0.550, 0.950)';

// Back
var easeInBack = exports.easeInBack = 'cubic-bezier(0.600, -0.280, 0.735, 0.045)';
var easeOutBack = exports.easeOutBack = 'cubic-bezier(0.175,  0.885, 0.320, 1.275)';
var easeInOutBack = exports.easeInOutBack = 'cubic-bezier(0.680, -0.550, 0.265, 1.550)';

exports.default = {
  easeInCubic: easeInCubic,
  easeOutCubic: easeOutCubic,
  easeInOutCubic: easeInOutCubic,
  easeInCirc: easeInCirc,
  easeOutCirc: easeOutCirc,
  easeInOutCirc: easeInOutCirc,
  easeInExpo: easeInExpo,
  easeOutExpo: easeOutExpo,
  easeInOutExpo: easeInOutExpo,
  easeInQuad: easeInQuad,
  easeOutQuad: easeOutQuad,
  easeInOutQuad: easeInOutQuad,
  easeInQuart: easeInQuart,
  easeOutQuart: easeOutQuart,
  easeInOutQuart: easeInOutQuart,
  easeInQuint: easeInQuint,
  easeOutQuint: easeOutQuint,
  easeInOutQuint: easeInOutQuint,
  easeInSine: easeInSine,
  easeOutSine: easeOutSine,
  easeInOutSine: easeInOutSine,
  easeInBack: easeInBack,
  easeOutBack: easeOutBack,
  easeInOutBack: easeInOutBack
};
